File: src/limitedwip/watchdog/components/WatchdogComponent.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package limitedwip.watchdog.components;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.components.AbstractProjectComponent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import limitedwip.common.LimitedWipCheckin;
import limitedwip.common.TimerComponent;
import limitedwip.common.settings.LimitedWIPSettings;
import limitedwip.common.settings.LimitedWipConfigurable;
import limitedwip.watchdog.Watchdog;

public class WatchdogComponent extends AbstractProjectComponent {
	private Watchdog watchdog;
	private TimerComponent timer;
	private IdeAdapter ideAdapter;


	public WatchdogComponent(Project project) {
		super(project);
		timer = ApplicationManager.getApplication().getComponent(TimerComponent.class);
	}

	@Override public void projectOpened() {
		LimitedWIPSettings settings = ServiceManager.getService(LimitedWIPSettings.class);
		final ChangeSizeCalculator changeSizeCalculator = new ChangeSizeCalculator(myProject);
		ideAdapter = new IdeAdapter(myProject, changeSizeCalculator);
		watchdog = new Watchdog(ideAdapter).init(convert(settings));

		timer.addListener(new TimerComponent.Listener() {
			@Override public void onUpdate(final int seconds) {
				ApplicationManager.getApplication().invokeLater(new Runnable() {
					@Override public void run() {
						// Project can be closed (disposed) during handover between timer thread and EDT.
						if (myProject.isDisposed()) return;
						watchdog.onTimer(seconds);
						changeSizeCalculator.onTimer();
					}
				}, ModalityState.any());
			}
		}, myProject);

		LimitedWipConfigurable.registerSettingsListener(myProject, new LimitedWipConfigurable.Listener() {
			@Override public void onSettingsUpdate(LimitedWIPSettings settings) {
				watchdog.onSettings(convert(settings));
			}
		});

		LimitedWipCheckin.Companion.registerListener(myProject, new LimitedWipCheckin.Listener() {
			@Override public void onSuccessfulCheckin(boolean allFileAreCommitted) {
				watchdog.onCommit();
			}
		});
	}

	public int currentChangeListSize() {
		return ideAdapter.currentChangeListSizeInLines().value;
	}

	public void toggleSkipNotificationsUntilCommit() {
        watchdog.toggleSkipNotificationsUntilCommit();
    }

	public void skipNotificationsUntilCommit(boolean value) {
		watchdog.skipNotificationsUntilCommit(value);
	}

	private static Watchdog.Settings convert(LimitedWIPSettings settings) {
		return new Watchdog.Settings(
				settings.watchdogEnabled,
				settings.maxLinesInChange,
				settings.notificationIntervalInSeconds(),
				settings.showRemainingChangesInToolbar
		);
	}
}

File: src/limitedwip/watchdog/components/WatchdogComponent.kt
package limitedwip.watchdog.components

import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import com.intellij.openapi.vcs.changes.Change
import limitedwip.common.TimerAppComponent
import limitedwip.common.settings.LimitedWipSettings
import limitedwip.common.settings.TimeUnit.Minutes
import limitedwip.common.toPathMatchers
import limitedwip.common.vcs.AllowCommit
import limitedwip.common.vcs.SuccessfulCheckin
import limitedwip.common.vcs.invokeLater
import limitedwip.common.vcs.registerSuccessfulCheckinListener
import limitedwip.watchdog.Watchdog
import limitedwip.watchdog.ui.WatchdogStatusBarWidget

class WatchdogComponentStartup : ProjectActivity {
    override suspend fun execute(project: Project) = WatchdogComponent(project).start()
}

class WatchdogComponent(private val project: Project) {
    @Volatile private var enabled = false

    fun start() {
        val settings = LimitedWipSettings.getInstance(project).toWatchdogSettings()
        val ide = WatchdogIde(project, ChangeSizeWatcher(project), WatchdogStatusBarWidget(project), settings)
        val watchdog = Watchdog(ide, settings)
        enabled = settings.enabled

        ide.listener = object : WatchdogIde.Listener {
            override fun onForceCommit() = watchdog.onForceCommit()
            override fun onSkipNotificationsUntilCommit() = watchdog.onSkipNotificationsUntilCommit()
            override fun onWidgetClick() = watchdog.toggleSkipNotificationsUntilCommit()
        }

        AllowCommit.addListener(project, object : AllowCommit.Listener {
            override fun allowCommit(project: Project, changes: List<Change>) =
                project != this@WatchdogComponent.project || watchdog.isCommitAllowed(ide.currentChangeListSizeInLines())
        })

        TimerAppComponent.getInstance().addListener(project, object : TimerAppComponent.Listener {
            override fun onUpdate() {
                // Optimisation to avoid scheduling task when component is not enabled.
                if (!enabled) return

                invokeLater(ModalityState.any()) {
                    // Project can be closed (disposed) during handover between timer thread and EDT.
                    if (project.isDisposed) return@invokeLater
                    watchdog.onTimer()
                }
            }
        })

        LimitedWipSettings.getInstance(project).addListener(project, object : LimitedWipSettings.Listener {
            override fun onUpdate(settings: LimitedWipSettings) {
                enabled = settings.watchdogEnabled
                watchdog.onSettingsUpdate(settings.toWatchdogSettings())
            }
        })

        registerSuccessfulCheckinListener(project, object : SuccessfulCheckin.Listener {
            override fun onSuccessfulCheckin(allChangesAreCommitted: Boolean) = watchdog.onSuccessfulCommit()
        })
    }

    private fun LimitedWipSettings.toWatchdogSettings() =
        Watchdog.Settings(
            watchdogEnabled,
            maxLinesInChange,
            Minutes.toSeconds(notificationIntervalInMinutes),
            showRemainingChangesInToolbar,
            noCommitsAboveThreshold,
            exclusions.toPathMatchers()
        )
}
